### PR TITLE
homework

### PR DESCRIPTION
**Pull request by zwtt1994**

### CODE REVIEW

**Comment by zwtt1994** [comment]
修改了思考题，加上了dnn-distribute

File: basic-tf/zhangweite/README.md [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+## 思考题
+1. 请使用自己的语言举例说明TensorFlow的声明式编程和我们日常的编程方式上的区别? 如何理解TensorFlow中一切都是计算图上的节点? TensorFlow中写日志操作也是计算图中的节点吗?  
+-TF的声明式编程相当于提前建立好计算框架，包括计算节点以及数据维度，在搭建框架时并不进行计算。在执行的时候给予输入以及执行命令，整个计算框架才开始计算。而日常的编程则是顺序执行的，一旦执行完语句之后就会得到结果。TF这种运行模式适合于机器学习训练以及预测的模式。  
+-TF中的计算图包括节点与连接线，节点表示数据输入/输出、权值以及计算模块，连接线则表示了数据的传递方向，所以TensorFlow计算图上的节点基本反应了整个计算框架。  
+-写日志操作只是记录节点输出，应该不算计算图中的节点。  
+
+2. 下面的代码有区别吗?请详细说明。 其中预定义 `W = tf.Variable(tf.zeros(10))`, `g = np.random.rand(10)`
+    - `W = W - 0.1 * g`
+    - `W.assign_sub(0.1 * g)`  
+原先W是Variable，第一行代码返回的是Tensor，第二行代码返回的是Variable。
```
Comment (tracholar) [inline basic-tf/zhangweite/README.md]: 除了返回值的区别，还有什么重要区别？

File: basic-tf/zhangweite/README.md [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+variable具有存储空间，在训练过程中属性保持不变，并且可以更新其数值。  
+tensor保存的是数据从输入到输出的一系列计算过程，并提供op可以run出当前状态的输出结果。  
+在计算类型中，能够用tensor类型计算的都可以利用variable类型替代。  
+在计算逻辑中，variable是最简单的tensor，没有任何计算过程。  
+操作（Session.run）是指执行某一部分计算图的过程。  
+
+4. 请使用tensorboard查看计算图可视化的结果, 思考: 为什么TensorFlow要建立一张反向计算图, 而不是直接在原图中实现BP算法?  
+TensorFlow的求导方式为自动微分方式（介于符号计算和数值计算之间），即每个最小计算单元（op）都定义好了forward和backward（grad）函数，整个计算图里的求导则按照链式法则计算。  
+自动微分具有前向和后向两种方式，当求dy/dx时（x的维度为m，y的维度为n），所求的结果是一个m*n的雅可比矩阵。借由微分的加法原则，求一次d/dx和一次d/dy都是一次矩阵乘法，所以前向计算方式需要计算m次自动微分，后向方式需要计算n次自动微分。  
+在神经网络中，深层的网络维度一般比浅层的要小，所以TensorFlow采用的是反向自动微分方式计算梯度，所以需要建立反向图。
```
Comment (tracholar) [inline basic-tf/zhangweite/README.md]: bp一定要建立反向图吗？？再查查资料

File: basic-tf/zhangweite/README.md [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+操作（Session.run）是指执行某一部分计算图的过程。  
+
+4. 请使用tensorboard查看计算图可视化的结果, 思考: 为什么TensorFlow要建立一张反向计算图, 而不是直接在原图中实现BP算法?  
+TensorFlow的求导方式为自动微分方式（介于符号计算和数值计算之间），即每个最小计算单元（op）都定义好了forward和backward（grad）函数，整个计算图里的求导则按照链式法则计算。  
+自动微分具有前向和后向两种方式，当求dy/dx时（x的维度为m，y的维度为n），所求的结果是一个m*n的雅可比矩阵。借由微分的加法原则，求一次d/dx和一次d/dy都是一次矩阵乘法，所以前向计算方式需要计算m次自动微分，后向方式需要计算n次自动微分。  
+在神经网络中，深层的网络维度一般比浅层的要小，所以TensorFlow采用的是反向自动微分方式计算梯度，所以需要建立反向图。  
+
+5. tensorboard的直方图怎么查看? 怎么将训练集和测试集的数据显示在同一张图上?  
+tensorboard --logdir=log，进入tensorboard选择直方图查看。  
+在tensorboard网页scalars页面左下角都钩上即可。
```
Comment (tracholar) [inline basic-tf/zhangweite/README.md]: 描述一下TensorFlow直方图的含义，跟平常见到的直方图的差异

File: basic-tf/zhangweite/README.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+checkpoint：保存模型文件的路径信息列表，可以查询到最新保存的文件信息  
+model.ckpt.meta：保存网络结构  
+model.ckpt.data：保存网络权值  
+model.ckpt.index：保存索引信息，作用是将Tensor name对应到Tensor的类型、形状、偏移和校验等信息，即关联meta文件和data文件。   
+
+7. 使用变量作用域的好处是啥?  
+tf中的作用域包括名称作用域和变量作用域，名称作用域中的variable和tensor都会加上作用域前缀。  
+为了共享变量，定义了tf.get_variable(‘name’,)，它无视名称作用域，若定义变量名称已经存在则会报错。  
+为了进一步管理tf.get_variable，定义了变量作用域，tf.get_variable在变量作用域中会加前缀。同时变量作用域还可以通过设定reuse=True来定义共享变量，并且定义了一个变量作用域后，相当于间接定义了一个名称作用域。此外，变量作用域还可以为tf.get_variable(‘name',)设置默认的初始化分布。  
+综上，变量作用域相当于加强版的名称作用域，可以更好的管理变量命名以及变量共享。
```
Comment (tracholar) [inline basic-tf/zhangweite/README.md]: 对计算图的可视化有什么帮助？

File: basic-tf/zhangweite/distribut-dnn.py [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+		# hook相当于时刻监听程序中的状态，对应的作出反应
+		# 这里指的是训练1000次后停止
+		hooks=[tf.train.StopAtStepHook(last_step=1000)]
+
+		# The MonitoredTrainingSession takes care of session initialization,
+		# restoring from a checkpoint, saving to a checkpoint, and closing when done
+		# or an error occurs.
+		# 监控训练过程，包括执行初始化、模型恢复和模型训练。
+		with tf.train.MonitoredTrainingSession(master=server.target,
+			is_chief=(FLAGS.task_index == 0),
```
Comment (tracholar) [inline basic-tf/zhangweite/distribut-dnn.py]: is_chief的作用是什么？

File: basic-tf/zhangweite/dnn.py [hunk 6]
```diff
@@ -0,0 +1,116 @@
...
+        conv_11 = tf.layers.conv2d( 
+            inputs=in_x, filters=8, kernel_size=3,strides=1, padding='same', activation=tf.nn.relu, name = 'c11'
+        )   #  (28, 28, 1) -> (28, 28, 8)
+        conv_12 = tf.layers.conv2d( 
+            inputs=conv_11, filters=16, kernel_size=3, strides=1, padding='same', activation=tf.nn.relu, name = 'c12'
+        )   #  (28, 28, 8) -> (28, 28, 16)
+        conv_13 = tf.layers.conv2d( 
+            inputs=conv_12, filters=32, kernel_size=5, strides=2, padding='same', activation=tf.nn.relu, name = 'c13'
+        )   #  (28, 28, 16) -> (14, 14, 32)
+        ld_14 = tf.layers.dropout(inputs = conv_13, rate=0.4, training=tf_is_train)
```
Comment (tracholar) [inline basic-tf/zhangweite/dnn.py]: 解释一下tf_is_train 对 dropout和batch normalization 的影响是什么？

Comment (tracholar) [reply basic-tf/zhangweite/dnn.py]: 写在实验报告中

File: basic-tf/zhangweite/distribut-dnn.py (hunk 5 above)
Comment (tracholar) [reply basic-tf/zhangweite/distribut-dnn.py]: 写在实验报告中

File: basic-tf/zhangweite/dnn.py (hunk 6 above)
Comment (tracholar) [reply basic-tf/zhangweite/dnn.py]: 并解释一下dropout和batch normalization你自己的理解

File: basic-tf/zhangweite/dnn.py [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+
+
+with tf.Session() as sess:
+    logdir = './tmp/log_mlp'
+    train_writer = tf.summary.FileWriter(logdir + '/train', sess.graph)
+    test_writer = tf.summary.FileWriter(logdir + '/test')
+
+    sess.run(tf.group(tf.global_variables_initializer(), tf.local_variables_initializer()))
+   
+    saver = tf.train.Saver(tf.get_collection(tf.GraphKeys.GLOBAL_VARIABLES))
```
Comment (tracholar) [inline basic-tf/zhangweite/dnn.py]: 解释一下 tf.GraphKeys 的作用？

File: basic-tf/zhangweite/dnn_dataset.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    iterator_y = dataset_y.make_one_shot_iterator()
+    tf_y_pre = iterator_y.get_next()
+    tf_y = tf.reshape(tf_y_pre, [-1,10]) 
+    tf_is_train = tf.placeholder(tf.bool, None)
+
+def mlp_net(x_):
+    with tf.name_scope('MLP'):
+        in_x = tf.reshape(x_, [-1,784]) 
+        l1 = tf.layers.dense(inputs = in_x, units = 512, activation = tf.nn.relu, name ='l1',
+            kernel_initializer = tf.initializers.random_normal(mean=0, stddev=1))
```
Comment (tracholar) [inline basic-tf/zhangweite/dnn_dataset.py]: 解释一下stddev取值的理由？这个值是最优的吗？

Comment (tracholar) [reply basic-tf/zhangweite/dnn_dataset.py]: 代码中的解释都写到实验报告中

File: basic-tf/zhangweite/dnn_dataset.py [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+
+with tf.name_scope('loss'):
+    Loss = tf.losses.softmax_cross_entropy(onehot_labels=tf_y, logits=output) 
+    corr = tf.equal(tf.argmax(output, 1), tf.argmax(tf_y,1))
+    acc = tf.reduce_mean(tf.cast(corr, tf.float32))
+
+step = tf.Variable(0, trainable=False)
+optimizer = tf.train.AdamOptimizer(learning_rate=1e-4).minimize(Loss, global_step=step)
+update_op = tf.get_collection(tf.GraphKeys.UPDATE_OPS)
```
Comment (tracholar) [inline basic-tf/zhangweite/dnn_dataset.py]: UPDATE_OPS 在你的代码中包含哪些操作？

**Comment by tracholar** [comment]
在深入思考一下

File: basic-tf/zhangweite/README.md [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+6.batch_normalization在该例子中效果并不是很好（batch=100）。  
+7.网络权值随机初始化能够加速网络收敛，甚至能和激活函数搭配使用（selu）。  
+8.test准确率不能直接调用tf.metrics.accuracy，它指的是整个session内，所有feed_dict数据的正确率。  
+## 实验结果
+batch_size=100的情况下，  
+-MLP训练10000个batch，test acc=0.9398  
+-CNN训练1000个batch，test acc=0.9832  
+-RNN训练10000个batch，test acc=0.9804  
+上述三个网络均采用了简单的网络结构（电脑跑不动），并且在训练中没有收敛。  
+总结来说，CNN网络能够较好的对图片做分类，但RNN能够在参数量较少的情况下达到较高的准确率令人惊讶。
```
[the remark this answers is not in the document]
Comment (tracholar) [reply basic-tf/zhangweite/README.md]: 可以给个解释么？

**Comment by tracholar** [comment]
@RbingChen @wenshiqi  你们两也可以一起讨论一下

File: basic-tf/zhangweite/README.md [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+
+7. 使用变量作用域的好处是啥?  
+更方便的管理参数名，比如两个作用域中的命名可以重复。  
+
+8. 分布式训练的时候, 程序是如何组织server端代码和worker端代码, 使得两部分代码可以放到一个脚本文件中的?  
+用if-elif来组织的。  
+
+9. 用自己的话简述分布式训练时Server和Worker分别做了哪些事情? 参数服务器与MapReduce的区别有哪些?  
+-Server负责更新参数，Worker负责计算梯度。  
+-参数服务器相较于MapReduce的优点：参数全局共享，支持异步通信，新增节点不需要重启。
```
[the remark this answers is not in the document]
Comment (tracholar) [reply basic-tf/zhangweite/README.md]: 具体来说，比如用PS实现逻辑回归和用MapReduce实现逻辑回归，之间有哪些具体差异？

File: basic-tf/zhangweite/distribut-dnn.py [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+			loss = tf.losses.softmax_cross_entropy(onehot_labels=tf_y, logits=output) 
+
+			global_step = tf.contrib.framework.get_or_create_global_step()
+
+			train_op = tf.train.AdagradOptimizer(0.01).minimize(loss, global_step=global_step)
+
+		# The StopAtStepHook handles stopping after running given steps.
+		# hook相当于时刻监听程序中的状态，对应的作出反应
+		# 这里指的是训练1000次后停止
+		hooks=[tf.train.StopAtStepHook(last_step=1000)]
```
Comment (tracholar) [inline basic-tf/zhangweite/distribut-dnn.py]: 解释TensorFlow分布式训练时，hook的作用原理

File: basic-tf/zhangweite/dnn.py [hunk 13]
```diff
@@ -0,0 +1,116 @@
...
+tf.summary.scalar('acc', acc)
+merged = tf.summary.merge_all()
+
+
+with tf.Session() as sess:
+    logdir = './tmp/log_mlp'
+    train_writer = tf.summary.FileWriter(logdir + '/train', sess.graph)
+    test_writer = tf.summary.FileWriter(logdir + '/test')
+
+    sess.run(tf.group(tf.global_variables_initializer(), tf.local_variables_initializer()))
```
Comment (tracholar) [inline basic-tf/zhangweite/dnn.py]: global_variables_initializer 和 local_variables_initializer 分别是初始化哪些变量？
